packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts

```diff
@@ -0,0 +1,484 @@
+   *
+   * An interface endpoint is an elastic network interface with a private IP
+   * address that serves as an entry point for traffic destined to a supported
+   * service.
+   */
+  Interface = 'Interface',
+
+  /**
+   * Gateway
+   *
+   * A gateway endpoint is a gateway that is a target for a specified route in
+   * your route table, used for traffic destined to a supported AWS service.
+   */
+  Gateway = 'Gateway'
+}
+
+/**
+ * A service for a gateway VPC endpoint.
+ */
+export interface IGatewayVpcEndpointService {
+  /**
+   * The name of the service.
+   */
+  readonly name: string;
+}
+
+/**
+ * An AWS service for a gateway VPC endpoint.
+ */
+export class GatewayVpcEndpointAwsService implements IGatewayVpcEndpointService {
+  public static readonly DynamoDb = new GatewayVpcEndpointAwsService('dynamodb');
+  public static readonly S3 = new GatewayVpcEndpointAwsService('s3');
+
+  /**
+   * The name of the service.
+   */
+  public readonly name: string;
+
+  constructor(name: string, prefix?: string) {
+    this.name = `${prefix || 'com.amazonaws'}.${cdk.Aws.region}.${name}`;
+  }
+}
+
+/**
+ * Options to add a gateway endpoint to a VPC.
+ */
+export interface GatewayVpcEndpointOptions {
+  /**
+   * The service to use for this gateway VPC endpoint.
+   */
+  readonly service: IGatewayVpcEndpointService;
+
+  /**
+   * Where to add endpoint routing.
+   *
+   * @default private subnets
+   */
+  readonly subnets?: SubnetSelection
+}
+
+/**
+ * Construction properties for a GatewayVpcEndpoint.
+ */
+export interface GatewayVpcEndpointProps extends GatewayVpcEndpointOptions {
+  /**
+   * The VPC network in which the gateway endpoint will be used.
+   */
+  readonly vpc: IVpcNetwork
+}
+
+/**
+ * A gateway VPC endpoint.
+ */
+export class GatewayVpcEndpoint extends VpcEndpoint implements IGatewayVpcEndpoint {
+  /**
+   * Imports an existing gateway VPC endpoint.
+   */
+  public static import(scope: cdk.Construct, id: string, props: GatewayVpcEndpointImportProps): IGatewayVpcEndpoint {
+    return new ImportedGatewayVpcEndpoint(scope, id, props);
+  }
+
+  /**
+   * The gateway VPC endpoint identifier.
+   */
+  public readonly vpcEndpointId: string;
+
+  /**
+   * The date and time the gateway VPC endpoint was created.
+   */
+  public readonly creationTimestamp: string;
+
+  constructor(scope: cdk.Construct, id: string, props: GatewayVpcEndpointProps) {
+    super(scope, id);
+
+    const subnets = props.vpc.selectSubnets(props.subnets);
+    const routeTableIds = (subnets as VpcSubnet[]).map(subnet => subnet.routeTableId);
+
+    const endpoint = new CfnVPCEndpoint(this, 'Resource', {
+      policyDocument: new cdk.Token(() => this.policyDocument),
+      routeTableIds,
+      serviceName: props.service.name,
+      vpcEndpointType: VpcEndpointType.Gateway,
+      vpcId: props.vpc.vpcId
+    });
+
+    this.vpcEndpointId = endpoint.vpcEndpointId;
+    this.creationTimestamp = endpoint.vpcEndpointCreationTimestamp;
```

Comment:
Please use the vpcEndpoint prefix. It hints that this is a deploy-time attribute (soon more formally)